src/backend/context.rs

```diff
@@ -0,0 +1,706 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use backend::*;
+use backend::cork_state::CorkState;
+use backend::var_array::VarArray;
+use capi::PULSE_OPS;
+use cubeb;
+use libc::strcmp;
+use libc::{c_char,c_void};
+use libpulse_sys::*;
+use semver;
+use std::default::Default;
+use std::ffi::CStr;
+use std::ptr;
+
+macro_rules! dup_str {
+    ($Dst: expr, $Src: expr) => {
+        if !$Dst.is_null() {
+            pa_xfree($Dst as *mut _);
+        }
+
+        $Dst = pa_xstrdup($Src);
+    }
+}
+
+const PA_RATE_MAX: u32 = 48000*8;
+
+fn pa_channel_to_cubeb_channel(channel: pa_channel_position_t) -> cubeb::Channel
+{
+    assert!(channel != PA_CHANNEL_POSITION_INVALID);
+    match channel {
+        PA_CHANNEL_POSITION_MONO => cubeb::Channel::Mono,
+        PA_CHANNEL_POSITION_FRONT_LEFT => cubeb::Channel::Left,
+        PA_CHANNEL_POSITION_FRONT_RIGHT => cubeb::Channel::Right,
+        PA_CHANNEL_POSITION_FRONT_CENTER => cubeb::Channel::Center,
+        PA_CHANNEL_POSITION_SIDE_LEFT => cubeb::Channel::LeftSurround,
+        PA_CHANNEL_POSITION_SIDE_RIGHT => cubeb::Channel::RightSurround,
+        PA_CHANNEL_POSITION_REAR_LEFT => cubeb::Channel::RearLeftSurround,
+        PA_CHANNEL_POSITION_REAR_CENTER => cubeb::Channel::RearCenter,
+        PA_CHANNEL_POSITION_REAR_RIGHT => cubeb::Channel::RearRightSurround,
+        PA_CHANNEL_POSITION_LFE => cubeb::Channel::LowFrequency,
+        _ => cubeb::Channel::Invalid
+    }
+}
+
+fn channel_map_to_layout(cm: &pa_channel_map) -> cubeb::ChannelLayout
+{
+    let mut cubeb_map: cubeb::ChannelMap = Default::default();
+    cubeb_map.channels = cm.channels as u32;
+    for i in 0usize..cm.channels as usize {
+        cubeb_map.map[i] = pa_channel_to_cubeb_channel(cm.map[i]);
+    }
+    unsafe { cubeb::cubeb_channel_map_to_layout(&cubeb_map) }
+}
+
+#[derive(Debug)]
+pub struct Context
+{
+    pub ops: *const cubeb::Ops,
+    pub mainloop: *mut pa_threaded_mainloop,
+    pub context: *mut pa_context,
+    pub default_sink_info: *mut pa_sink_info,
+    pub context_name: *const i8,
+    pub collection_changed_callback: cubeb::DeviceCollectionChangedCallback,
+    pub collection_changed_user_ptr: *mut c_void,
+    pub error: bool,
+    pub version_2_0_0: bool,
+    pub version_0_9_8: bool
+}
+
+destroy!(Context);
+
+impl Context
+{
+    pub fn new(name: *const i8) -> Result<Box<Self>>
+    {
+        let mut ctx = Box::new(Context{
+            ops: &PULSE_OPS,
+            mainloop: pa_threaded_mainloop_new(),
+            context: 0 as *mut _,
+            default_sink_info: 0 as *mut _,
+            context_name: name,
+            collection_changed_callback: None,
+            collection_changed_user_ptr: 0 as *mut _,
+            error: true,
+            version_0_9_8: false,
+            version_2_0_0: false,
+        });
+
+        pa_threaded_mainloop_start(ctx.mainloop);
+
+        if ctx.pulse_context_init() != cubeb::OK {
+            ctx.destroy();
```

Comment:
It probably is, but to aid the review, I didn't want to change too much of the structure of the C->rust code in this pass.